skorch/helper.py

```diff
@@ -131,6 +132,105 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
 
+# This class must be an instance of Sequence and have an ndim
+# attribute because sklearn will test this.
+class SliceDataset(Sequence):
+    # pylint: disable=anomalous-backslash-in-string
+    """Helper class that wraps a torch dataset to make it work with
+    sklearn.
+
+    Sometimes, sklearn will touch the input data, e.g. when splitting
+    the data for a grid search. This will fail when the input data is
+    a torch dataset. To prevent this, use this wrapper class for your
+    dataset.
+
+    Note: This class will only return the X value by default (i.e. the
+    first value returned by indexing the original dataset). Sklearn,
+    and hence skorch, always require 2 values, X and y. Therefore, you
+    still need to provide the y data separately.
+
+    Note: This class behaves similarly to a PyTorch
+    :class:`~torch.utils.data.Subset` when it is indexed by a slice or
+    numpy array: It will return another ``SliceDataset`` that
+    references the subset instead of the actual values. Only when it
+    is indexed by an int does it return the actual values. The reason
+    for this is to avoid loading all data into memory when sklearn,
+    for instance, creates a train/validation split on the
+    dataset. Data will only be loaded in batches during the fit loop.
+
+    Examples
+    --------
+    >>> X = MyCustomDataset()
+    >>> search = GridSearchCV(net, params, ...)
+    >>> search.fit(X, y)  # raises error
+    >>> ds = SliceDataset(X)
+    >>> search.fit(ds, y)  # works
+
+    Parameters
+    ----------
+    dataset : torch.utils.data.Dataset
+      A valid torch dataset.
+
+    n : int (default=0)
```

Comment:
Is there a better variable name for this?

Reply:
I have no good idea, any suggestions?

Reply:
How about renaming the parameter to `key` to be in line with builtins like `sorted` and `max`?

Reply:
The argument doesn't serve the same purpose, though, since `key` in those functions is a callable.

A more similar function would be `itemgetter`, where the argument is called `item`. But `itemgetter` also has `items` for returning a tuple, so now I wonder if we should just use `itemgetter` under the hood to make it possible to return multiple items.